Topic/vern/script inlining

vpax opened the pull request:
This branch fixes merge issues with my previous attempt.  It supersedes both the stand-alone AST duplication and the previous inlining branches.

rsmmr commented:
This is now just one big commit, I though we wanted to split this up into at least two parts (the earlier AST duplication, then inlining); ideally even more.

jsiwek commented on src/Frame.h:
```diff
@@ -316,6 +325,13 @@ class Frame :  public Obj {
+	 * This is how we support inlined functions without having to
+	 * alter the offsets associated with their local variables.
+	 */
+	int current_offset;
```
Was hard to tell whether there's some missing applications of this new offset.

Example: calling `SetElement(n, v)` and then `GetElement(n)` seems like it may not actually return a reference to the `v` that was just set at index `n` when offset is non-zero.

Idea for improvement: wrap the `frame` member in a struct w/ overloaded `operator[]` that ensures the offset gets applied to all accesses?

vpax replied:
I'm pretty sure there aren't any missing applications, as they'd fail immediately (running the test suite with inlining will present zillions of such opportunities).  The one you flagged is technically an issue but in practice not, as the only time `GetElement` is called with an numeric index rather than an `ID*` is during function call invocation, and those can't occur with non-zero `current_offset`.

If you want this to be tidier, eems simpler to make that call stand out a bit (e.g., `GetElementUsingIndex`) and adjust it to add in the offset.

jsiwek commented on src/reporter.bif:
```diff
@@ -22,7 +22,7 @@ module Reporter;
 function Reporter::info%(msg: string%): bool
 	%{
-	reporter->PushLocation(frame->GetCall()->GetLocationInfo());
+	reporter->PushLocation(frame->GetCall()->Original()->GetLocationInfo());
```
Might it become error prone to remember or confusing to know when to do `Original()->GetLocationInfo()` vs. `GetLocationInfo()` ?  Could/should it work in a more generalized way such that asking an object for location info defaults to returning the original location (if there is one) ?

vpax replied:
Do you mean make `GetLocationInfo()` virtual, and override it for those classes (`Expr`, `Stmt`) that have the notion of `Original()`?  I agree that that's cleaner.  It might require adding some logic to get the non-Original location in some circumstances, but that's likely better isolated.

jsiwek reviewed (commented):
@vpax made a first pass of review/revisions, please see if I messed up anything important for later on.

Also, a couple tests fail.  Run:

```
btest -d -a AST-dup bifs/to_addr.zeek bifs/to_double_from_string.zeek
```

These were also failing before I made changes, but I think btest/btest-diff had just not been taught how to do new alternative/baseline tricks and so didn't report them correctly.  However, running `zeek` directly on those two would have showed the difference between `ZEEK_DUPLICATE_ASTS=1` and the standard config.

Lastly, what's the idea for the general state of btest alternatives like `AST-dup` and `inline` in the end?  How many total alternatives?

Updating CI to cover those two seems warranted before merging, but not sure we should change the default test targets to also run them such that developers are bothered by the 3+ local test suite runs.

jsiwek commented on src/script_opt/ScriptOpt.cc:
```diff
@@ -0,0 +1,65 @@
+void analyze_func(ScriptFuncPtr f)
+	{
+	auto info = new FuncInfo(f, {NewRef{}, f->GetScope()}, f->CurrentBody());
+	funcs.push_back(info);
```
Went ahead and did that change.

vpax commented on src/script_opt/Inline.cc:
```diff
@@ -0,0 +1,242 @@
+			}
+		}
+
+	// Functions that are candidates for inlining.
+	std::unordered_set<FuncInfo*> candidates;
```
No, that's vestigial and can be removed.

vpax commented:
> made a first pass of review/revisions, please see if I messed up anything important for later on.

@jsiwek I've now looked these over, they all seem fine -  thanks!

vpax commented:
> btest -d -a AST-dup bifs/to_addr.zeek bifs/to_double_from_string.zeek

These reflect the same issue with `Original()` vs. `GetLocationInfo()` as discussed above.  I believe if we got with making the latter virtual, then that will address these differences.  Make sense?

vpax commented:
> Lastly, what's the idea for the general state of btest alternatives like `AST-dup` and `inline` in the end? How many total alternatives?
> 
> Updating CI to cover those two seems warranted before merging, but not sure we should change the default test targets to also run them such that developers are bothered by the 3+ local test suite runs.

My thinking was that (1) we'd keep the alternatives (and there will be several more as we add additional elements of script optimization), as they're handy for localizing problems when debugging, but (2) we wouldn't include them in CI, since they incur a lot of work for what's nearly always an unsurprising answer.  Does that work?

jsiwek commented:
> My thinking was that (1) we'd keep the alternatives (and there will be several more as we add additional elements of script optimization), as they're handy for localizing problems when debugging, but (2) we wouldn't include them in CI, since they incur a lot of work for what's nearly always an unsurprising answer. Does that work?

Yeah, thanks.

> > btest -d -a AST-dup bifs/to_addr.zeek bifs/to_double_from_string.zeek
> 
> These reflect the same issue with `Original()` vs. `GetLocationInfo()` as discussed above. I believe if we got with making the latter virtual, then that will address these differences. Make sense?

Yes, would you mind taking a crack at the idea of making it virtual ?  I.e. just push a commit into this branch and I can review/massage afterward.

vpax commented:
@jsiwek just pushed the `GetLocationInfo()`-as-virtual approach.  Definitely better, cleaned up the code a tad and addressed the test suite issues.  I also looked over your recent pushes, all good.

jsiwek commented:
@vpax thanks, that was the last loose end I had so did the merge
